src/librustc/ty/util.rs

```diff
@@ -1090,14 +1090,18 @@ fn needs_drop_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>
         // Can refer to a type which may drop.
         // FIXME(eddyb) check this against a ParamEnv.
         ty::Dynamic(..)
-        | ty::Projection(..)
         | ty::Param(_)
         | ty::Bound(..)
         | ty::Placeholder(..)
         | ty::Opaque(..)
         | ty::Infer(_)
         | ty::Error => true,
 
+        ty::Projection(..) => {
+            let normalized = tcx.normalize_erasing_regions(param_env, ty);
+            if let ty::Projection(..) = normalized.kind { true } else { needs_drop(normalized) }
```

Comment:
IIRC the approach `ty::layout` takes is to check `normalized == ty`, but this is just nit-picking.